stan/math/opencl/kernels/normal_id_glm_lpdf.hpp

```diff
@@ -0,0 +1,161 @@
+#ifndef STAN_MATH_OPENCL_KERNELS_NORMAL_ID_GLM_LPDF_HPP
+#define STAN_MATH_OPENCL_KERNELS_NORMAL_ID_GLM_LPDF_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/opencl/kernel_cl.hpp>
+
+namespace stan {
+namespace math {
+namespace opencl_kernels {
+
+// \cond
+static const char* normal_id_glm_kernel_code = STRINGIFY(
+    // \endcond
+    /**
+     * GPU implementation of Generalized Linear Model (GLM)
+     * with Normal distribution and identity link function.
+     *
+     * Must be run with at least N threads and local size equal to LOCAL_SIZE_.
+     * @param[in] y vector parameter
+     * @param[in] x design matrix
+     * @param[in] alpha intercept (in log odds)
+     * @param[in] beta weight vector
+     * @param[in] sigma_glob (Sequence of) scale parameters for the normal
+     * @param[out] mu_derivative_glob intermediate variable used in the model
+     * @param[out] mu_derivative_sum partially summed mu_derivative_glob (1
+     * value per work group)
+     * @param[out] y_minus_mu_over_sigma_squared_sum Partial sum of squares of
+     * y, scaled by expected mean and given variance.
+     * @param[out] sigma_derivative derivative with respect to sigma
+     * @param[out] log_sigma_sum partially summed logarithm of sigma (1 value
+     * per work group)
+     * @param N number of cases
+     * @param M number of attributes
+     * @param is_alpha_vector 0 or 1 - whether alpha is a vector (alternatively
+     * it is a scalar)
+     * @param is_sigma_vector 0 or 1 - whether sigma is a vector (alternatively
+     * it is a scalar)
+     * @param need_mu_derivative interpreted as boolean - whether mu_derivative
+     * needs to be computed
+     * @param need_mu_derivative_sum interpreted as boolean - whether
+     * mu_derivative_sum needs to be computed
+     * @param need_sigma_derivative interpreted as boolean - whether
+     * sigma_derivative needs to be computed
+     * @param need_log_sigma_sum interpreted as boolean - whether log_sigma_sum
+     * needs to be computed
+     */
+    __kernel void normal_id_glm(
+        __global double* mu_derivative_glob, __global double* mu_derivative_sum,
+        __global double* y_minus_mu_over_sigma_squared_sum,
+        __global double* sigma_derivative, __global double* log_sigma_sum,
+        const __global double* y, const __global double* x,
+        const __global double* alpha, const __global double* beta,
+        const __global double* sigma_glob, const int N, const int M,
+        const int is_alpha_vector, const int is_sigma_vector,
+        const int need_mu_derivative, const int need_mu_derivative_sum,
+        const int need_sigma_derivative, const int need_log_sigma_sum) {
```

Comment:
This is pretty hard to review with so many paths in one function. We should think of a nicer way to do this

Reply:
If we refactor this kernel into multiple kernels we suffer performance penalty for loading data multiple times. If we refactor it into multiple functions within the same kernel we still have to pass all these arguments to the kernel. 

Maybe we could organize the arguments into some structs, but I don't see a nice way to do it.

Reply:
Yeah I recognize that, though I think the main performance we would lose is some function invocations and a few places where we can reuse local memory. Would it be significant for you to break this up into smaller kernels? If the performance penalty is ballpark less than 5% than I would feel a lot more comfortable with future maintanence

Reply:
I'm pretty sure I can construct a case where penalty would be close to 2 for splitting this into two kernels (low number of attributes, large number of cases), so I would not want to go that way.

I hope we will eventually be able to use kernel generator to replace these complicated GLM kernels, but we are not there yet.

Reply:
@SteveBronder The thing is these kernels are memory transfer bound and we are dealing with execution times at the level of sub miliseconds even for large inputs. 

For stuff like cholesky, multiply, mdivide etc, it doesnt really make a difference if you extract something out because execution times are a few orders of magnitude larger there. Here its actually a huge difference. Its also why Tadej is working on the kernel generator.

Reply:
Alright I'll trust yinz judgement here let's just make sure we doc this all well